### PR TITLE
[WIP] Handling subprocesses for app hooks

### PR DESCRIPTION
**Pull request by Himavanth**
### Current behaviour of app hooks 
- subprocess launched with npm run-script
- subprocess inherits same io streams as parent process
- SIGINT shuts down the parent process too
- > so any subprocess that uses SIGINT for exit cannot do so without closing the parent process down
- > this also introduces ambiguity at the stream level resulting in issues like the tty display bug
	

### Solution
- Let the subprocess exit without closing the parent process
- > we could remove the SIGINT listeners from parent process temporarily for this
- Use different IO streams for parent and child processes to remove any possibility of stream takeovers/closures

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/lib/app-helper.js [hunk 1]
```diff
@@ -51,8 +51,37 @@ async function runPackageScript (scriptName, dir, cmdArgs = []) {
...
+    let termListeners = process.listeners('SIGINT')
+    process.removeAllListeners('SIGINT')
+    process.on('SIGINT', () => {
+      if(!childPro.killed) {
+        childPro.kill()
+      }
+    })
+
+    const childPro = execa('npm', ['run-script', scriptName].concat(cmdArgs), { cwd: dir, stdio: 'pipe' })
+    if(!childPro) return
```
Comment (moritzraho) [inline src/lib/app-helper.js:64]: missing braces

File: src/lib/app-helper.js [hunk 2]
```diff
@@ -51,8 +51,37 @@ async function runPackageScript (scriptName, dir, cmdArgs = []) {
...
   const pkg = await fs.readJSON(path.join(dir, 'package.json'))
   if (pkg && pkg.scripts && pkg.scripts[scriptName]) {
-    return execa('npm', ['run-script', scriptName].concat(cmdArgs), { cwd: dir, stdio: 'inherit' })
-  } else {
+
+    let termListeners = process.listeners('SIGINT')
+    process.removeAllListeners('SIGINT')
+    process.on('SIGINT', () => {
+      if(!childPro.killed) {
+        childPro.kill()
```
Comment (moritzraho) [inline src/lib/app-helper.js:59]: shouldn't sigint also kill the main process ? seems like the user would need to CTRL +C two times

File: src/lib/app-helper.js [hunk 3]
```diff
@@ -51,8 +51,37 @@ async function runPackageScript (scriptName, dir, cmdArgs = []) {
   aioLogger.debug(`running npm run-script ${scriptName} in dir: ${dir}`)
   const pkg = await fs.readJSON(path.join(dir, 'package.json'))
   if (pkg && pkg.scripts && pkg.scripts[scriptName]) {
-    return execa('npm', ['run-script', scriptName].concat(cmdArgs), { cwd: dir, stdio: 'inherit' })
-  } else {
+
+    let termListeners = process.listeners('SIGINT')
+    process.removeAllListeners('SIGINT')
```
Comment (shazron) [inline src/lib/app-helper.js:56]: why the side-effect of removing all listeners for the event? it's possible some script/code/hook might be listening as well for some reason

File: src/lib/app-helper.js [hunk 4]
```diff
@@ -51,8 +51,37 @@ async function runPackageScript (scriptName, dir, cmdArgs = []) {
...
+    try {
+      await childPro
+    }catch(err){
+      // Ignore
+    }
+    process.stdin.unpipe()
+    process.stdin.resume()
+    // Adding the SIGINT listeners back
+    if(termListeners.length>0) {
+      process.removeAllListeners('SIGINT')
```
Comment (shazron) [inline src/lib/app-helper.js:78]: similar to my comment above

**Comment by moritzraho** [comment]
One potential solution (not decided) as discussed this morning with the team:

- have a flag or additional hooks (e.g. `post-app-run-detached`) to treat long running processes as detached and cleanup on sigint (like the openwhisk jar in `--local`) instead of the main process waiting for the child process to terminate (which would be the case for short running child processes) 

we probably need to expose a distinction between short and long running commands in hooks:
- short running hooks might resolve a dependency on the run command (e.g. moving some files pre-built) and hence the main process needs to wait for the process to terminate
- long running hooks like the `asset-compute` command that spawns the NUI devtool UI need to run detached and be cleaned up on termination

**Comment by jdelbick** [comment]
@moritzraho regarding your comment:

> long running hooks like the asset-compute command that spawns the NUI devtool UI need to run detached and be cleaned up on termination. Those could be started and terminated in the same way as the openwhisk jar subprocess in --local

what changes would we need to make in the devtool to support this?

**Comment by moritzraho** [comment]
@jdelbick that's all hypothetical, but let's say we expose an additional `<pre/post>-app-run-detached` hook then the only thing you would need to do is use that hook to run the long running process instead of the current `<pre/post>-app-run`

**Pull request closed without merging**